### PR TITLE
stm32f0 / thumbinator integration

### PR DESCRIPTION
**Pull request by kbeckmann**

### CODE REVIEW

**Comment by kbeckmann** [comment]
Bunch of dirty stuff here and a few // and #if 0 leftovers to clean up. But it works!

File: Makefile [hunk 1]
```diff
@@ -4,83 +4,73 @@ CC      := $(CROSS_COMPILE)gcc
...
-	SDK/Drivers/STM32L4xx_HAL_Driver/Src/stm32l4xx_hal_uart.c \
-	SDK/Drivers/STM32L4xx_HAL_Driver/Src/stm32l4xx_hal_uart_ex.c \
-	SDK/Drivers/STM32L4xx_HAL_Driver/Src/stm32l4xx_ll_usb.c \
+TARGET ?= xildebug_big
+
+# Include target first since it chooses a platform
+include targets/$(TARGET)/build.mk
+include platforms/$(PLATFORM)/build.mk
+
+# Feature flags go here
```
Comment (EnJens) [inline Makefile]: shouldn't feature flags be target specific?

File: Makefile [hunk 2]
```diff
@@ -4,83 +4,73 @@ CC      := $(CROSS_COMPILE)gcc
...
+FEAT_POWER_PROFILER		?= 0
```
Comment (EnJens) [inline Makefile]: indentation seems off?

File: Makefile [hunk 3]
```diff
@@ -4,83 +4,73 @@ CC      := $(CROSS_COMPILE)gcc
...
+	ifeq ($(FEAT_DUT_SWITCH),0)
+	$(error FEAT_POWER_PROFILER requires FEAT_DUT_SWITCH)
+	endif
+TARGET_SRCS += \
+	app/drivers/mcp4018t.c \
+	app/power.c
+endif
+APP_CFLAGS  += -DFEAT_POWER_PROFILER=$(FEAT_POWER_PROFILER)
+
+ifeq ($(FEAT_DUT_SWITCH),1)
```
Comment (EnJens) [inline Makefile]: `ifeq ($(FEAT_DUT_SWITCH),max14662)` perhaps?

File: Makefile [hunk 4]
```diff
@@ -4,83 +4,73 @@ CC      := $(CROSS_COMPILE)gcc
...
+	$(PLATFORM_SRCS) \
+
+TARGET_INCLUDES += \
+	targets \
+	targets/$(TARGET) \
+	platforms \
+	platforms/$(PLATFORM) \
+	$(PLATFORM_INCLUDES) \
+
+MIDDLEWARE_SRCS += \
```
Comment (EnJens) [inline Makefile]: Maybe switch to an out-of-SDK freertos and just use portable's from SDK. More portable if we want to port to another chip

File: Makefile [hunk 5]
```diff
@@ -118,7 +108,7 @@ stlink:
 	@openocd \
 	-f interface/stlink-v2.cfg \
 	-c "transport select hla_swd" \
-	-f target/stm32l4x.cfg \
+	-f target/stm32f0x.cfg \
```
Comment (EnJens) [inline Makefile]: Maybe use something like OPENOCD_TARGET_CFG?

File: app/cdc_uart_bridge.c [hunk 6]
```diff
@@ -58,7 +57,8 @@ static void rx_task(void *p_arg)
 				;
 		}
 
-		r = uart_tx(rx_queue_item.data, rx_queue_item.len, portMAX_DELAY, true);
+		usb_cdc_tx(rx_queue_item.data, rx_queue_item.len);
```
Comment (EnJens) [inline app/cdc_uart_bridge.c]: so.... mirror all data back?

File: app/platform/usb/hid.h [hunk 7]
```diff
@@ -19,4 +19,4 @@
 
 err_t usb_hid_recv(struct usb_rx_queue_item *p_rx_queue_item, uint32_t timeout_ticks);
 err_t usb_hid_send(uint8_t *p_buf, uint16_t len);
-err_t usb_hid_init(USBD_HandleTypeDef *p_usbd, PCD_HandleTypeDef *p_pcd);
+err_t usb_hid_init(const void *p_data);
```
Comment (EnJens) [inline app/platform/usb/hid.h]: perhaps use a target-specific typedef'ed value instead? e.g. `const platform_usb_pdata_t *p_data;`
with that as
`typedef USBD_HandleTypeDef platform_usb_pdata_t;`

File: targets/xildebug_big/target.c [hunk 8]
```diff
@@ -0,0 +1,6 @@
+#include "platform/platform.h"
+
+err_t target_init(void)
+{
+	return platform_init();
+}
```
Comment (EnJens) [inline targets/xildebug_big/target.c]: missing newline

File: Makefile (hunk 1 above)
Comment (kbeckmann) [reply Makefile]: They are, they are just defined here. The implementations in app/ are platform independent.

File: Makefile (hunk 3 above)
Comment (kbeckmann) [reply Makefile]: Sure. There is a lot more we could decouple so you can have different switch implementations on different targets.

File: Makefile (hunk 4 above)
Comment (kbeckmann) [reply Makefile]: Yeah the code is essentially the same

File: Makefile (hunk 5 above)
Comment (kbeckmann) [reply Makefile]: Sure

File: app/cdc_uart_bridge.c (hunk 6 above)
Comment (kbeckmann) [reply app/cdc_uart_bridge.c]: As I wrote before it’s a bit rough patch right now. Need to clean up stuff like this and some #if 0

File: app/platform/usb/hid.h (hunk 7 above)
Comment (kbeckmann) [reply app/platform/usb/hid.h]: Mm like the gpio_impl.h I suppose, so each platform gets its own init function prototype.

File: Makefile (hunk 1 above)
Comment (kbeckmann) [reply Makefile]: I mean the available features in the codebase are defined here, they are turned on/off in target/$target/build.mk

File: Makefile (hunk 2 above)
Comment (kbeckmann) [reply Makefile]: tab = 4 spaces, might be that you view it without the glorious 4tab extension

**Comment by kbeckmann** [comment]
Tested on thumbinator and xildebug_big, and it seems to work great :)

File: Makefile (hunk 1 above)
Comment (EnJens) [reply Makefile]: True, I missed the `?=`

File: Makefile (hunk 4 above)
Comment (EnJens) [reply Makefile]: Is that a "nah, seems pointless" or?

**Comment by kbeckmann** [comment]
Uhhh what am I doing wrong in the Makefile (last few commits). Works fine locally but not on Travis. For some reason, the CFLAGS for xildebug_big are always there, then the actual target's cflags are appended.

**Pull request merged**